test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua

```diff
@@ -167,8 +167,7 @@ g.test_simple = function()
     }
     tx_step(g.server, 'tx2', "s:replace{2, 2}", diff)
     tx_gc(g.server, 100, nil)
-    local err = g.server:eval('return tx2:commit()')
-    assert(not err[1])
+    t.assert_equals(g.server:eval('return tx2:commit()'), '')
```

Comment:
`box.commit()` actually returns zero return values and our console encodes it as `'---\n...\n'`. The reason that it is interpreted as an empty string is `yaml.decode()` in `tnx_proxy`. I think it is better to handle this case explicitly in `txn_proxy` and return no values.

There are some ideas around this console peculiar (see #3714), but I doubt that we'll change it. At least not unconditionally, because a script may lean on it.

Reply:
> I think it is better to handle this case explicitly in txn_proxy and return no values.

I don't get how this can be applied in the context of this patch, could you please elabroate?

Reply:
```diff
diff --git a/test/box/lua/txn_proxy.lua b/test/box/lua/txn_proxy.lua
index 15b0e4add..02c6c7e19 100644
--- a/test/box/lua/txn_proxy.lua
+++ b/test/box/lua/txn_proxy.lua
@@ -22,7 +22,11 @@ local array_mt = { __serialize = 'array' }
 local mt = {
     __call = function(self, code_str)
         self.c1:put(code_str)
-        local res = yaml.decode(self.c2:get())
+        local res_str = self.c2:get()
+        if res_str == '---\n...\n' then
+            return
+        end
+        local res = yaml.decode(res_str)
         return type(res) == 'table' and setmetatable(res, array_mt) or res
     end,
     __index = {
```

It'll require updating of many usages, so deserves its own commit. But as result we'll get more straightforward behavior of `txn_proxy`.

Reply:
Alternatively we can update the testing code step-by-step:

```lua
diff --git a/test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua b/test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua
index df404d54c..49dda1577 100644
--- a/test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua
+++ b/test/box-luatest/gh_6150_memtx_tx_memory_monitoring_test.lua
@@ -69,6 +69,7 @@ g.before_each(function()
     g.server:start()
 
     g.server:eval('txn_proxy = require("test.box.lua.txn_proxy")')
+    g.server:eval('txn_proxy.cfg({decode_empty_document_as_no_values = true})')
     g.server:eval('s = box.schema.space.create("test")')
     g.server:eval('s:create_index("pk")')
     -- Clear txm before test
@@ -167,8 +168,7 @@ g.test_simple = function()
     }
     tx_step(g.server, 'tx2', "s:replace{2, 2}", diff)
     tx_gc(g.server, 100, nil)
-    local err = g.server:eval('return tx2:commit()')
-    assert(not err[1])
+    t.assert_is(g.server:eval('return tx2:commit()'), nil)
     diff = {
         ["txn"] = {
             ["statements"] = {
@@ -194,8 +194,7 @@ g.test_simple = function()
         }
     }
     tx_gc(g.server, 10, diff)
-    err = g.server:eval('return tx1:commit()')
-    assert(not err[1])
+    t.assert_is(g.server:eval('return tx1:commit()', nil))
     diff = {
         ["txn"] = {
             ["statements"] = {
@@ -421,8 +420,7 @@ g.test_user_data = function()
         },
     }
     tx_step(g.server, 'tx', 'ffi.C.box_txn_alloc(' .. alloc_size .. ')', diff)
-    local err = g.server:eval('return tx:commit()')
-    assert(not err[1])
+    t.assert_is(g.server:eval('return tx:commit()', nil))
     diff = {
         ["txn"] = {
             ["user"] = {
diff --git a/test/box/lua/txn_proxy.lua b/test/box/lua/txn_proxy.lua
index 15b0e4add..e42d95d91 100644
--- a/test/box/lua/txn_proxy.lua
+++ b/test/box/lua/txn_proxy.lua
@@ -19,10 +19,24 @@ local console = require('console')
 
 local array_mt = { __serialize = 'array' }
 
+local cfg = {
+    -- Disabled by default for compatibility.
+    --
+    -- When disabled returns an empty string for
+    -- a command that doesn't return any value.
+    decode_empty_document_as_no_values = false,
+}
+
 local mt = {
     __call = function(self, code_str)
         self.c1:put(code_str)
-        local res = yaml.decode(self.c2:get())
+        local res_str = self.c2:get()
+        if cfg.decode_empty_document_as_no_values then
+            if res_str == '---\n...\n' then
+                return
+            end
+        end
+        local res = yaml.decode(res_str)
         return type(res) == 'table' and setmetatable(res, array_mt) or res
     end,
     __index = {
@@ -51,4 +65,15 @@ local function new_txn_proxy()
     }, mt)
 end
 
-return { new = new_txn_proxy }
+local function cfg_txn_proxy(new_cfg)
+    for k, v in pairs(new_cfg) do
+        assert(cfg[k] ~= nil)
+        assert(type(v) == 'boolean')
+        cfg[k] = v
+    end
+end
+
+return {
+    new = new_txn_proxy,
+    cfg = cfg_txn_proxy,
+}
```

However it looks overcomplicated.

Reply:
I understand what you mean and I thought about this before too, but here I wanted to emphasize is that it is not applicable in scope of this patch: I suggest we open a separate ticket concerning this matter.

Reply:
> ```
>     g.server:eval('tx1:begin()')
>     g.server:eval('tx2:begin()')
>     g.server:eval('tx1("s:get(1)")')
>     g.server:eval('tx2("s:replace{1, 2}")')
>     g.server:eval("box.internal.memtx_tx_gc(10)")
> ```

The test is screwed: as long as it uses `eval` like that, replacing plain `assert` with `luatest.assert` is no better than putting lipstick on a pig. I think we should rewrite it using `exec` while we're at it or not touch it at all.

Reply:
I think it is a kind of thing, which only has a chance to be made if someone wants it. Not really requires an issue and not really worth it. I don't mind ignoring it.

Reply:
> Not really requires an issue and not really worth it.

No, I really believe it's also about code health: `txn_proxy` should ideally behave like streams and builtin operations.